src/middleware/lollipopMiddleware.ts

```diff
@@ -0,0 +1,131 @@
+import { pipe } from "fp-ts/lib/function";
+import * as O from "fp-ts/lib/Option";
+import * as TE from "fp-ts/lib/TaskEither";
+import * as T from "fp-ts/lib/Task";
+import * as B from "fp-ts/lib/boolean";
+import * as E from "fp-ts/lib/Either";
+import * as jose from "jose";
+import { Request, Response } from "express-serve-static-core";
+import { verifySignatureHeader } from "@mattrglobal/http-signatures";
+import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
+import { getPublicKey } from "../persistence/lollipop";
+import { ioDevServerConfig } from "../config";
+import { signAlgorithmToVerifierMap } from "../utils/httpSignature";
+import { serverUrl } from "../utils/server";
+import { getProblemJson } from "../payloads/error";
+
+type LollipopHTTPStatusError = {
+  code: number;
+  problemJson: string;
+};
+
+export const DEFAULT_LOLLIPOP_HASH_ALGORITHM = "sha256";
+
+export const isLollipopConfigEnabled = () =>
+  ioDevServerConfig.features.lollipop.enabled;
+
+export const lollipopMiddleware =
+  (nextMiddleware: (embeddedRequest: Request, _: Response) => void) =>
+  async (request: Request, response: Response) => {
+    pipe(
+      isLollipopConfigEnabled(),
+      B.fold(
+        () => nextMiddleware(request, response),
+        async () =>
+          pipe(
+            await verifyLollipopSignatureHeader(request, response),
+            E.foldW(
+              error => response.status(error.code).send(error.problemJson),
+              _ => nextMiddleware(request, response)
+            )
+          )
+      )
+    );
+  };
```

Comment:
Here we could avoid `async/await` with a more `fp-ts` idiomization... 

```suggestion
export const lollipopMiddleware =
  (nextMiddleware: (embeddedRequest: Request, _: Response) => void) =>
  (request: Request, response: Response) => {
    pipe(
      isLollipopConfigEnabled(),
      B.fold(
        () => nextMiddleware(request, response),
        () =>
          pipe(
            TE.tryCatch(
              () => verifyLollipopSignatureHeader(request, response),
              _ => _ as Error
            ),
            TE.map(verificationResult =>
              pipe(
                verificationResult,
                E.foldW(
                  error => response.status(error.code).send(error.problemJson),
                  _ => nextMiddleware(request, response)
                )
              )
            )
          )()
      )
    );
  };
```
Is a bit more verbose, as usual... 😊. What do you think?

Reply:
Yeah I prefer the one you are suggesting, let's adopt it